internal/handler/msg_saslstart.go

```diff
@@ -129,3 +153,116 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
 
 	return string(authcid), string(passwd), nil
 }
+
+// scramCredentialLookup looks up an user's credentials in the database.
+func (h *Handler) scramCredentialLookup(ctx context.Context, username, dbName string) (scram.StoredCredentials, error) {
+	adminDB, err := h.b.Database("admin")
+	if err != nil {
+		return scram.StoredCredentials{}, lazyerrors.Error(err)
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return scram.StoredCredentials{}, lazyerrors.Error(err)
+	}
+
+	var filter *types.Document
+
+	filter, err = usersInfoFilter(false, false, "", []usersInfoPair{
+		{username: username, db: dbName},
+	})
+	if err != nil {
+		return scram.StoredCredentials{}, lazyerrors.Error(err)
+	}
+
+	// Filter isn't being passed to the query as we are filtering after retrieving all data
+	// from the database due to limitations of the internal/backends filters.
+	qr, err := usersCol.Query(ctx, nil)
+	if err != nil {
+		return scram.StoredCredentials{}, lazyerrors.Error(err)
+	}
+
+	defer qr.Iter.Close()
+
+	for {
+		_, v, err := qr.Iter.Next()
+
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return scram.StoredCredentials{}, lazyerrors.Error(err)
+		}
+
+		matches, err := common.FilterDocument(v, filter)
+		if err != nil {
+			return scram.StoredCredentials{}, lazyerrors.Error(err)
+		}
+
+		if matches {
+			credentials := must.NotFail(v.Get("credentials")).(*types.Document)
+
+			if !credentials.Has("SCRAM-SHA-256") {
+				return scram.StoredCredentials{}, handlererrors.NewCommandErrorMsgWithArgument(
+					handlererrors.ErrMechanismUnavailable,
+					"User has no SCRAM-SHA-256 based authentication credentials registered",
+					"SCRAM-SHA-256",
+				)
+			}
+
+			cred := must.NotFail(credentials.Get("SCRAM-SHA-256")).(*types.Document)
+
+			salt := must.NotFail(base64.StdEncoding.DecodeString(must.NotFail(cred.Get("salt")).(string)))
+			storedKey := must.NotFail(base64.StdEncoding.DecodeString(must.NotFail(cred.Get("storedKey")).(string)))
+			serverKey := must.NotFail(base64.StdEncoding.DecodeString(must.NotFail(cred.Get("serverKey")).(string)))
+
+			return scram.StoredCredentials{
+				KeyFactors: scram.KeyFactors{
+					Salt:  string(salt),
+					Iters: int(must.NotFail(cred.Get("iterationCount")).(int32)),
+				},
+				StoredKey: storedKey,
+				ServerKey: serverKey,
+			}, nil
+		}
+	}
+
+	return scram.StoredCredentials{}, errors.New("user not found")
+}
+
+// saslStartSCRAMSHA256 extracts the initial challenge and respond the client.
```

Comment:
I don't quite understand the godoc line, so I suggest this (please check if it's correct):
```suggestion
// saslStartSCRAMSHA256 extracts the initial challenge, checks credentials, and returns the response that needs to be sent to the client.
//
// If credentials are not found, it returns XYZ.
```

Reply:
Probably

```
// saslStartSCRAMSHA256 extracts the initial challenge and attempts to move the
// authentication conversation forward.
```

(similar to how `*scram.ServerConversation.Step` explains) is more precise as it checks if the challenge can be resolved, rather than the credential itself.